### PR TITLE
Retrieve scale files from registry

### PR DESCRIPTION
**Pull request by SuperManifolds**
Previously known as pulldown (https://github.com/loopholelabs/scale/pull/33)

### CODE REVIEW

**Comment by jimmyaxod** [comment]
Thanks for taking this over @SuperManifolds

File: go/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	}
+}
+
+func WithCacheDirectory(cacheDirectory string) Option {
+	return func(config *Config) {
+		config.cacheDirectory = cacheDirectory
+	}
+}
+
+func WithApiKey(apiKey string) Option {
```
Comment (ShivanshVij) [inline go/registry/registry.go:66]: `WithAPIKey(apiKey string)`

File: go/registry/registry.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+		config.organization = organization
+	}
+}
+
+// Create a new runtime for a specific scalefile
+func New(function string, tag string, opts ...Option) (*scalefunc.ScaleFunc, error) {
+	// Default config
+	conf := &Config{
+		pullPolicy:     AlwaysPullPolicy,
+		cacheDirectory: "~/.cache/scale/functions",
```
Comment (ShivanshVij) [inline go/registry/registry.go:89]: save this in a const or use the homedir lib (https://pkg.go.dev/github.com/mitchellh/go-homedir) to get this?

File: go/registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+		apiKey:         "",
+		apiBaseUrl:     "https://api.scale.sh/v1",
```
Comment (ShivanshVij) [inline go/registry/registry.go:91]: const as well

File: go/registry/registry.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+	// Decode to a scalefile
+	err = sf.Decode(data)
+	if err != nil {
+		return nil, err
+	}
+	return sf, nil
+}
+
+// build a filename from the config
+// TODO: We should use the hash in the filename for optimization
```
Comment (ShivanshVij) [inline go/registry/registry.go:152]: we should probably do this now rather than later

File: go/registry/registry.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+	// Try to read the scalefile
+	path := fmt.Sprintf("%s%c%s", conf.cacheDirectory, os.PathSeparator, f)
+	return scalefunc.Read(path)
+}
+
+// Save a scalefile to our local cache
+func saveToCache(function string, tag string, conf *Config, data []byte) error {
+	err := os.MkdirAll(conf.cacheDirectory, os.ModePerm)
+	if err != nil {
+		return err
```
Comment (ShivanshVij) [inline go/registry/registry.go:170]: Check if already exists

Comment (ShivanshVij) [reply go/registry/registry.go:170]: os.IsExist or something

File: go/registry/registry.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+	return nil
+}
+
+func removeCache(conf *Config) error {
+	return os.RemoveAll(conf.cacheDirectory)
+}
+
+// PulldownResponse API response when requesting a function
+type GetFunctionResponse struct {
```
Comment (ShivanshVij) [inline go/registry/registry.go:200]: We should probably use a swagger-generated client if we can here, but this is fine for now

File: go/registry/registry.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+	Name         string `json:"name"`
+	Tag          string `json:"tag"`
+	Organization string `json:"organization"`
+	Public       bool   `json:"public"`
+	Hash         string `json:"hash"`
+	PresignedURL string `json:"presigned_url"`
+}
+
+func apiRequest(function string, tag string, conf *Config) (*GetFunctionResponse, error) {
+	client := &http.Client{}
```
Comment (ShivanshVij) [inline go/registry/registry.go:210]: Let's use http.DefaultClient here

File: go/registry/registry.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+	if res.StatusCode != 200 {
+		bodyBytes, err := io.ReadAll(res.Body)
+		if err != nil {
+			return nil, err
+		}
+		bodyString := string(bodyBytes)
+		return nil, errors.New(bodyString)
+	}
+
+	response := &GetFunctionResponse{}
```
Comment (ShivanshVij) [inline go/registry/registry.go:233]: response := new(GetFunctionResponse)

File: go/registry/registry_test.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+func TestRegistryDownload(t *testing.T) {
+	/* This test requires a valid API key for the scale dev api to run, the SCALE_API_KEY environment variable must be
+	set in the testing environment. */
+	apiKey := os.Getenv("SCALE_API_KEY")
+	if apiKey == "" {
+		t.Skip("Skipping test, SCALE_API_KEY environment variable not set")
+	}
+	sf, err := New("TestRegistryDownload", "1",
+		WithApiKey(apiKey),
+		WithBaseUrl("https://api.dev.scale.sh/v1"),
```
Comment (ShivanshVij) [inline go/registry/registry_test.go:81]: Const or set via env as well?

File: ts/src/registry/index.test.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    if (apiKey == undefined) {
+        console.log("SCALE_API_KEY not set, skipping test")
+        return
+    }
+
+    const scaleFunc = await New(
+        "TestRegistryDownload",
+        "1",
+        WithApiKey(apiKey),
+        WithApiBaseUrl("https://api.dev.scale.sh/v1"),
```
Comment (ShivanshVij) [inline ts/src/registry/index.test.ts:68]: const or set via env too?

File: ts/src/registry/index.ts [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+export function WithOrganization(organization: string): Option {
+    return (c: Config) => {
+        c.organization = organization;
+    }
+}
+
+export async function New(func: string, tag: string, ...opts: Option[]): Promise<ScaleFunc> {
+  const config: Config = {
+    pullPolicy: AlwaysPullPolicy,
+    cacheDirectory: "~/.cache/scale/functions",
```
Comment (ShivanshVij) [inline ts/src/registry/index.ts:75]: Some way to use the homedir lib here too? The `~` won't work on some operating systems.

File: ts/src/registry/index.ts [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+    apiKey: "",
+    apiBaseUrl: "https://api.scale.sh/v1",
```
Comment (ShivanshVij) [inline ts/src/registry/index.ts:77]: const

File: ts/src/registry/index.ts [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+  if (hash !== response.hash) {
+    throw ErrHasMismatch;
+  }
+
+  scaleFunc = ScaleFunc.Decode(new Uint8Array(data));
+  saveToCache(func, tag, config, scaleFunc);
+  return scaleFunc;
+}
+
+function buildFilename(func: string, tag: string, config: Config): string {
```
Comment (ShivanshVij) [inline ts/src/registry/index.ts:108]: Hash in the filename here too

File: ts/src/registry/index.ts [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+      response.on('data', (chunk) => {
+        body.push(chunk);
+      })
+
+      response.on('end', () => {
+        const data = Buffer.concat(body).toString();
+        resolve(JSON.parse(data));
+      })
+
+      response.on('error', (error) => {
```
Comment (ShivanshVij) [inline ts/src/registry/index.ts:170]: Does this trigger if the response code isn't 200?

**Review by ShivanshVij — CHANGES_REQUESTED** [review]
Looks good but needs some changes

**Pull request merged**